### PR TITLE
Checksum fix

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

*Description of changes:*

- Deprecate old S3 checksum related API
- Introduced the new checksum Config API and new features in there.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/software/amazon/awssdk/crt/s3/ChecksumAlgorithm.java [hunk 1]
```diff
@@ -45,4 +50,27 @@ private static Map<Integer, ChecksumAlgorithm> buildEnumMapping() {
...
     private static Map<Integer, ChecksumAlgorithm> enumMapping = buildEnumMapping();
+
+    /**
+     * @hidden Marshals a list of algorithm into an array for Jni to deal with
+     *
+     * @param algorithms list of algorithms
+     * @return a int[] that with the [algorithms.nativeValue, *]
+     */
+    public static int[] marshallAlgorithmsForJNI(final List<ChecksumAlgorithm> algorithms) {
+        if (algorithms == null) {
```
Comment (DmitriyMusatkin) [inline src/main/java/software/amazon/awssdk/crt/s3/ChecksumAlgorithm.java:61]: style suggestion: this whole method can be something like
Optional.ofNullable(algorithms).map(ChecksumAlgorithm::getNativeValue).toArray().orElseDefault(null);

File: src/main/java/software/amazon/awssdk/crt/s3/ChecksumConfig.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    private ChecksumLocation location = ChecksumLocation.NONE;
+    private ChecksumAlgorithm checksumAlgorithm = ChecksumAlgorithm.NONE;
+    private boolean validateChecksum = false;
+    private List<ChecksumAlgorithm> validateChecksumAlgorithmList = null;
+
+    public ChecksumConfig() {
+    }
+
+    /**
+     * If NONE. No request payload checksum will be add and calculated.
```
Comment (DmitriyMusatkin) [inline src/main/java/software/amazon/awssdk/crt/s3/ChecksumConfig.java:38]: nit: will be added*

File: src/main/java/software/amazon/awssdk/crt/s3/ChecksumConfig.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+     *
+     * If HEADER, the checksum will be calculated by client and added related header
```
Comment (DmitriyMusatkin) [inline src/main/java/software/amazon/awssdk/crt/s3/ChecksumConfig.java:40]: nit: and related header added*

File: src/main/java/software/amazon/awssdk/crt/s3/ChecksumConfig.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+     */
+    public ChecksumAlgorithm getChecksumAlgorithm() {
+        return this.checksumAlgorithm;
+    }
+
+    /**
+     * Enable checksum mode header will be attached to get requests, this will tell
+     * s3 to send back checksums headers if they exist.
+     *
+     * For object has checksum for the whole object, the checksum of whole object
```
Comment (DmitriyMusatkin) [inline src/main/java/software/amazon/awssdk/crt/s3/ChecksumConfig.java:91]: nit: sounds a bit awkward. probably want "for object that has checksum..."

File: src/native/s3_client.c [hunk 5]
```diff
@@ -523,10 +525,28 @@ JNIEXPORT jlong JNICALL Java_software_amazon_awssdk_crt_s3_S3Client_s3ClientMake
         resume_token = aws_jni_byte_cursor_from_jbyteArray_acquire(env, jni_resume_token);
     }
 
+    struct aws_s3_checksum_config checksum_config = {
+        .location = checksum_location,
+        .checksum_algorithm = checksum_algorithm,
+        .validate_response_checksum = validate_response,
+    };
+
+    struct aws_array_list response_checksum_list;
```
Comment (DmitriyMusatkin) [inline src/native/s3_client.c:534]: you always clean it up in done, but there are already jumps into done before this declaration. cleaning up undeclared variable will be a segfault

File: src/main/java/software/amazon/awssdk/crt/s3/ChecksumAlgorithm.java (hunk 1 above)
Comment (TingDaoK) [reply src/main/java/software/amazon/awssdk/crt/s3/ChecksumAlgorithm.java]: did it a bit different by adding another function to convert the list to an int array. Thanks for the suggestion!

**Review by DmitriyMusatkin — APPROVED** [review]

**Pull request merged**